sklearn/datasets/openml.py

```diff
@@ -242,11 +242,10 @@ def _convert_arff_data(arff_data, col_slice_x, col_slice_y, shape=None):
             count = -1
         else:
             count = shape[0] * shape[1]
-        data = np.fromiter(itertools.chain.from_iterable(arff_data),
-                           dtype='float64', count=count)
+        data = np.array(list(itertools.chain.from_iterable(arff_data)))
```

Comment:
I think that this change will reverse the enhancement from #13312.
@jorisvandenbossche Do you know what would be best here. It seems that we make our own `fromiter` which will select the column of X and y when iterating over the iterator. We should know the size in advance as well, isn't it?

Reply:
Yes, this reverses the recent enhancement, to use fromiter. Thinking about it, this change is somewhat messier with fromiter. Not sure what the right solution is. The purpose of fromiter was to avoid materialising all the python objects, which was consuming a lot of temporary memory. But there are other ways to chunk the reading, I suppose.

The bigger problem here is that we're temporarily converting all numeric values to strings, which undoes the conversion work in the arff library.

Reply:
I suspect that we will need to treat string columns separately before this, with an operation like:
```py
string_col_idxs = ...  # a list of string column indices
if string_col_idxs:
    numeric_slices = [slice(start, stop) for start, stop in zip([None] + string_col_idxs, string_col_idxs + [None])]
    arff_data = (row[sl] for row in arff_data for sl in numeric_slices)
```

Though this may all be essentially replicating functionality in Pandas, so if we just supported pandas output, we could avoid this.